### PR TITLE
chore: bump deps, implement unit tests

### PR DESCRIPTION
**Pull request by Unix4ever**
Implement unit-tests for the controller.
Bump dependencies.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -18,8 +18,8 @@ GO_LDFLAGS += -s -w
 
 ARTIFACTS := _out
 
-TOOLS ?= ghcr.io/siderolabs/tools:v1.3.0-1-g712379c
-PKGS ?= v1.3.0
+TOOLS ?= ghcr.io/siderolabs/tools:v1.4.0
```
Comment (smira) [inline Makefile:21]: we have that updated with new Go 1.20, better to take values from 1.4.0 of Talos `Makefile`

File: go.mod [hunk 2]
```diff
@@ -1,142 +1,149 @@
...
-	sigs.k8s.io/cluster-api v1.3.1
-	sigs.k8s.io/controller-runtime v0.13.1
+	k8s.io/api v0.26.1
+	k8s.io/apiextensions-apiserver v0.26.1
+	k8s.io/apimachinery v0.26.1
+	k8s.io/apiserver v0.26.1
+	k8s.io/client-go v0.26.1
+	k8s.io/klog/v2 v2.90.1
+	k8s.io/utils v0.0.0-20230209194617-a36077c30491
+	sigs.k8s.io/cluster-api v1.4.1
```
Comment (smira) [inline go.mod:30]: we should update controller-gen/conversion-gen versions as well  - cluster-api repo has version, or you can take a look at CABPT

File: controllers/configs.go [hunk 3]
```diff
@@ -48,6 +48,17 @@ func (r *TalosControlPlaneReconciler) talosconfigForMachines(ctx context.Context
 	}
 
 	clusterName := tcp.GetLabels()["cluster.x-k8s.io/cluster-name"]
+	if clusterName == "" {
```
Comment (smira) [inline controllers/configs.go:51]: this looks like something which is already in cluster-api:

```go
// GetOwnerCluster returns the Cluster object owning the current resource.
func GetOwnerCluster(ctx context.Context, c client.Client, obj metav1.ObjectMeta) (*clusterv1.Cluster, error) {
```

File: controllers/controllers_test.go [hunk 4]
```diff
@@ -0,0 +1,918 @@
...
+			Name:      "coredns",
+			Namespace: namespace.Name,
+		},
+		Data: map[string]string{
+			"Corefile": "original-core-file",
+		},
+	}
+	g.Expect(fakeClient.Create(ctx, corednsCM)).To(Succeed())
+
+	kubeadmCM := &corev1.ConfigMap{
```
Comment (smira) [inline controllers/controllers_test.go:416]: what is that???

Comment (Unix4ever) [reply controllers/controllers_test.go:416]: copy-paste issue ;) This PR is still WIP. Should have marked it as draft, debugging `unit-tests` step in CI

File: controllers/configs.go (hunk 3 above)
Comment (Unix4ever) [reply controllers/configs.go:51]: it doesn't seem equivalent to me. In our code we read the cluster name from the label.

File: Makefile (hunk 1 above)
Comment (Unix4ever) [reply Makefile]: ok, used the same from `v1.4.0` tag of Talos.

**Review by frezbo — APPROVED** [review]

File: controllers/controllers_test.go [hunk 5]
```diff
@@ -0,0 +1,847 @@
...
+	// - be less than 64 characters long.
+	// - be an empty string OR consist of alphanumeric characters, '-', '_' or '.'.
+	// - start and end with an alphanumeric character
+	if len(validation.IsValidLabelValue(str)) == 0 {
+		return str
+	}
+	hasher := fnv.New32a()
+	_, err := hasher.Write([]byte(str))
+	if err != nil {
+		// At time of writing the implementation of fnv's Write function can never return an error.
```
[the remark this answers is not in the document]
Comment (Unix4ever) [reply controllers/controllers_test.go:692]: yes, copied and reworked.

File: controllers/configs.go (hunk 3 above)
Comment (smira) [reply controllers/configs.go:51]: anyways we should use proper way - see that function. not sure if we need our own ways vs. what Cluster API does

Comment (Unix4ever) [reply controllers/configs.go]: gotcha, fixed

File: controllers/controllers_test.go [hunk 6]
```diff
@@ -0,0 +1,847 @@
...
+	got := r.ClusterToTalosControlPlane(cluster)
+	g.Expect(got).To(BeNil())
+}
+
+func TestReconcilePaused(t *testing.T) {
+	g := NewWithT(t)
+
+	clusterName := "foo"
+
+	// Test: cluster is paused and kcp is not
```
Comment (smira) [inline controllers/controllers_test.go:190]:
```suggestion
	// Test: cluster is paused and tcp is not
```

File: controllers/controllers_test.go [hunk 7]
```diff
@@ -0,0 +1,847 @@
...
+var (
+	ctx            = ctrl.SetupSignalHandler()
+	fakeScheme     = runtime.NewScheme()
+	ms             *machineService
+	machineAddress string
+	secretsBundle  *generate.SecretsBundle
+)
+
+//nolint:govet
+type machineService struct {
```
Comment (smira) [inline controllers/controllers_test.go:72]: this is more of a nit, so please take it as a suggestion

I try to split parts of the test infra into separate files to make tests themselves more readable.

e.g. this `machineService` and some of the code at the bottom of the file might make a better fit in separate files, so the tests are easier to find and follow, and as time goes, we know where to put things

File: controllers/controllers_test.go [hunk 8]
```diff
@@ -0,0 +1,847 @@
...
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	controlplanev1 "github.com/siderolabs/cluster-api-control-plane-provider-talos/api/v1alpha3"
+)
+
+var (
```
Comment (smira) [inline controllers/controllers_test.go:63]: are these variables good as globals? this feels a bit weird to me. each test _might_ probably run in parallel, in any order?

**Review by smira — APPROVED** [review]

File: controllers/controllers_test.go (hunk 8 above)
Comment (Unix4ever) [reply controllers/controllers_test.go]: Used a suite instead of globals.

File: controllers/controllers_test.go (hunk 7 above)
Comment (Unix4ever) [reply controllers/controllers_test.go]: done

**Pull request merged**